Add thick mode for oracledb (#14166)
DataDog/integrations-core: oracle/README.md

#### Prerequisite

> To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client, or download the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If not using the JDBC method, the minimum [supported version][2] is Oracle 12c.

Reviewer: ```suggestion
To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client via `python-oracledb` Thick mode, or download the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If not using the JDBC method, the minimum [supported version][2] is Oracle 12c.
```
Just adding this note since you mention `Thick mode` later in this README but I don't think we mention what `Thick mode` refers to.

> To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client via `python-oracledb` Thick mode, or download the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If not using the JDBC method, the minimum [supported version][2] is Oracle 12c.

Reviewer 2: ```suggestion
To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client through `python-oracledb` thick mode, or download the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If you are not using the JDBC method, the minimum [supported version][2] is Oracle 12c.
```

> To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client by using `python-oracledb` thick mode, or download the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If you are not using the JDBC method, the minimum [supported version][2] is Oracle 12c.

Reviewer 3: ```suggestion
To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client, or the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If you are not using the JDBC method, the minimum [supported version][2] is Oracle 12c.
```

To use the Oracle integration you can either use the native client (no additional install steps required), the Oracle Instant Client, or download the Oracle JDBC driver (Linux only). To use the Oracle integration with JDBC, download the Oracle JDBC driver. If not using the JDBC method, the minimum [supported version][2] is Oracle 12c.
Due to licensing restrictions, the JDBC library is not included in the Datadog Agent, but can be downloaded directly from Oracle.

> With Agent v7.42.x, the Agent no longer requires or supports installing the Instant Client libraries. If you're on an older version of the Agent and want to use the Instant Client, refer to the [Oracle Instant Client][3] setup instructions.

Reviewer: Can we also get rid of the link at `[3]`?

Author: That's paragraph is not included anymore because we do support it now

> 

Reviewer 3: should we provide here some instructions on how to install the instant client?

Author: I actually have a link mentioned in the troubleshooting section. If that doesn't suffice, I can re-add one in a more obvious location.

Reviewer 3: We need installation instructions on the installation section, same as we have for JDBC driver

Author: I added the instructions from v7.41.x that we removed.

sqlplus <USER>/<PASSWORD>@(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCPS)(HOST=<HOST>)(PORT=<PORT>))(SERVICE_NAME=<SERVICE_NAME>)))
```

> *Note*: When using the [Oracle Instant Client][16] connection, move three files to the `network/admin` directory of the client libraries used by your application:

Reviewer: I'm a bit confused about this `network/admin` directory. Would this be the same regardless of OS? Did we need to do this back when we supported the instant client with the old library?

Author: Nope we didn't need to do this with the old client but we did need to use the `TNS_ADMIN` environment variable, however it was never explicitly mentioned. We can use the same variable but in the case it's not set, I just mentioned where python-oracle looks by default.
> Since the tnsnames.ora and sqlnet.ora files are not in the default location, your application needs to indicate where they are, either with the config_dir parameter to [oracledb.init_oracle_client()](https://python-oracledb.readthedocs.io/en/latest/api_manual/module.html#oracledb.init_oracle_client), or using the TNS_ADMIN environment variable. See [Optional Oracle Net Configuration Files](https://python-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#optnetfiles). (Neither of these settings are needed, and you do not need to edit sqlnet.ora, if you have put all the files in the network/admin directory.)

https://python-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#install-the-wallet-and-network-configuration-files

Reviewer 2: ```suggestion
When using the [Oracle Instant Client][16] connection, move three files to the `network/admin` directory of the client libraries used by your application:
```

  * `tnsnames.ora`: Maps net service names used for application connection strings to your database services.
  * `sqlnet.ora`: Configures Oracle Network settings.

>   * `cwallet.sso`: Enables SSL or TLS connections in thick mode. Keep this file secure.

Reviewer: I saw that in [this comment](https://github.com/DataDog/integrations-core/pull/14166#discussion_r1145129203), we remove reference to the python-oracledb library thick mode, although I think if we mention "thick mode" in the README, we should still mention python-oracledb since that term is specific to the library. Another option is we don't mention thick mode and be library agnostic in the README. WDYT @hithwen?

>   * `cwallet.sso`: Enables SSL or TLS connections in thick mode. Keep this file secure.

Reviewer 3: We need to stick to one way or the other, either we say "native client", "instant client" or "thick mode" but let's use the same terminology everywhere

>   * `tnsnames.ora` - Maps net service names used for application connection strings to your database services
>   * `sqlnet.ora` - Configures Oracle Network settings
>   * `cwallet.sso` - Enables SSL/TLS connections in Thick mode. Keep this file secure

Reviewer 2: ```suggestion
  * `tnsnames.ora`: Maps net service names used for application connection strings to your database services.
  * `sqlnet.ora`: Configures Oracle Network settings.
  * `cwallet.sso`: Enables SSL or TLS connections in thick mode. Keep this file secure.
```

## Troubleshooting

### Common problems

#### Oracle Native Client
- If you encounter a `DPY-6000: cannot connect to database`:
  ```text
  Failed to connect to Oracle DB, error: DPY-6000: cannot connect to database. Listener refused connection. (Similar to ORA-12660)
  ```
    - Ensure Native Network Encryption or Checksumming is not enabled. If they are enabled, you must use thick mode by setting `use_instant_client: true`.

*Note:* See our documentation [here][3] on Oracle Instant Client setup.

#### Oracle Instant Client
- Verify that both the Oracle Instant Client and SDK files are located in the same directory.
The structure of the directory should look similar:
  ```text
  |____sdk/

>   |____network/

Author: I think it makes sense to reinforce visually even though we do have a note added about it. I'll add it. 🙇

## Troubleshooting

### Common problems

#### Oracle Native Client
- If you encounter a `DPY-6000: cannot connect to database`:
  ```text
  Failed to connect to Oracle DB, error: DPY-6000: cannot connect to database. Listener refused connection. (Similar to ORA-12660)
  ```

>     - Ensure Native Network Encryption or Checksumming is not enabled. If they are enabled, you must use thick mode by setting `use_instant_client: true`.

Reviewer 2: ```suggestion
 - Ensure Native Network Encryption or Checksumming are not enabled. If they are enabled, you must use thick mode by setting `use_instant_client: true`.
```

> *Note:* See our documentation [here][3] on Oracle Instant Client setup.

Reviewer 2: ```suggestion
For more information about setting up the Oracle Instant Client, see the [Oracle integration documentation][3].
```

#### Oracle Instant Client

> - Verify that both the Oracle Instant Client and SDK files are located in the same directory.

Reviewer: I see that in the `python-oracledb` [docs](https://python-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#setting-the-oracle-client-library-directory), we have different directions to initialize a Thick mode connection on different OS's. Would it be a good idea to add an optional config option so users can specify their instant client location?
